From 6cf6552e03450692d038c36e899ee13e5de7ced2 Mon Sep 17 00:00:00 2001
From: Thomas Wu <quic_wthomas@quicinc.com>
Date: Thu, 12 Sep 2024 17:44:30 -0700
Subject: [PATCH] wifi: ath12k: Correct FILS Discovery for Non-TX BSS

Currently code sends a FILS enable/disable WMI command
for each BSS that is available.

In a MBSS scenario, driver will send a FILS WMI Enable
for TX BSS VAP and a FILS WMI disable for each Non-TX
BSS VAP that is configured.

Due to sending a disable for the Non-TX BSS VAPs, this will
completely turn off the FILS Discovery feature in firmware.

To use the FILS Discovery feature properly, driver must
send FILS WMI commands to TX BSS VAPS only.

Signed-off-by: Thomas Wu <quic_wthomas@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -4649,10 +4649,19 @@ static int ath12k_mac_fils_discovery(str
 	struct ath12k *ar = arvif->ar;
 	struct sk_buff *tmpl;
 	struct ath12k_hw *ah = ar->ah;
+	struct ieee80211_bss_conf *link_conf;
 	int ret;
 	u32 interval;
 	bool unsol_bcast_probe_resp_enabled = false;
 
+	rcu_read_lock();
+	link_conf = ath12k_get_link_bss_conf(arvif);
+	if (link_conf && link_conf->nontransmitted) {
+		rcu_read_unlock();
+		return 0;
+	}
+	rcu_read_unlock();
+
 	if (info->fils_discovery.max_interval) {
 		interval = info->fils_discovery.max_interval;
 
